cargo-auditable/src/collect_audit_data.rs

```diff
@@ -4,13 +4,33 @@ use std::str::from_utf8;
 
 use crate::{
     auditable_from_metadata::encode_audit_data, cargo_arguments::CargoArgs,
-    rustc_arguments::RustcArgs,
+    rustc_arguments::RustcArgs, sbom_precursor,
 };
 
 /// Calls `cargo metadata` to obtain the dependency tree, serializes it to JSON and compresses it
 pub fn compressed_dependency_list(rustc_args: &RustcArgs, target_triple: &str) -> Vec<u8> {
-    let metadata = get_metadata(rustc_args, target_triple);
-    let version_info = encode_audit_data(&metadata).unwrap();
+    let sbom_path = std::env::var_os("CARGO_SBOM_PATH");
+
+    // If cargo has created precursor SBOM files, use them instead of `cargo metadata`.
+    let version_info = if sbom_path.as_ref().map(|p| !p.is_empty()).unwrap_or(false) {
+        let sbom_paths = std::env::split_paths(&sbom_path.unwrap()).collect::<Vec<_>>();
+        // Cargo may create multiple SBOM precursor files.
+        // We can't control per-binary (or cdylib) dependency information, just grab the first non-rlib SBOM we find.
+        let sbom_path = sbom_paths
+            .iter()
+            .find(|p| !p.ends_with(".rlib.cargo-sbom.json"))
+            .unwrap_or_else(|| &sbom_paths[0]);
+        let sbom_data: Vec<u8> = std::fs::read(sbom_path)
+            .unwrap_or_else(|_| panic!("Failed to read SBOM file at {}", sbom_path.display()));
```

Comment:
`path.display()` requires Rust 1.87; I would prefer to avoid bumping MSRV so aggressively. `.to_string_lossy()` is fast enough and works on older rustc versions.

Reply:
https://doc.rust-lang.org/std/path/struct.Path.html#method.display, it's been around since Rust 1.0.0?